webknossos-datastore/app/com/scalableminds/webknossos/datastore/services/DataSourceService.scala

```diff
@@ -86,6 +88,75 @@ class DataSourceService @Inject()(
     } yield ()
   }
 
+  private def sendRealPaths(dataSources: List[InboxDataSource]) =
+    for {
+      _ <- Fox.successful(())
+      magPathBoxes = dataSources.map(ds => (ds, determineMagRealPathsForDataSource(ds)))
+      pathInfos = magPathBoxes.map {
+        case (ds, Full(magPaths)) => DataSourcePathInfo(ds.id, magPaths)
+        case (ds, Failure(msg, _, _)) =>
+          logger.error(s"Failed to determine real paths for mags of $ds: $msg")
+          DataSourcePathInfo(ds.id, List())
+        case (ds, Empty) =>
+          logger.error(s"Failed to determine real paths for mags of $ds")
+          DataSourcePathInfo(ds.id, List())
+      }
+      _ <- dataSourceRepository.publishRealPaths(pathInfos)
+    } yield ()
+
+  private def determineMagRealPathsForDataSource(dataSource: InboxDataSource) = tryo {
+    val organizationPath = dataBaseDir.resolve(dataSource.id.organizationId)
+    val datasetPath = organizationPath.resolve(dataSource.id.directoryName)
+    dataSource.toUsable match {
+      case Some(usableDataSource) => {
+        usableDataSource.dataLayers.flatMap { dataLayer =>
+          val rawLayerPath = datasetPath.resolve(dataLayer.name)
+          val absoluteLayerPath = if (Files.isSymbolicLink(rawLayerPath)) {
+            resolveRelativePath(datasetPath, Files.readSymbolicLink(rawLayerPath))
+          } else {
+            rawLayerPath.toAbsolutePath
+          }
+          dataLayer.mags.map { mag =>
+            val (magURI, isRemote) = getMagURI(datasetPath, absoluteLayerPath, mag)
+            if (isRemote) {
+              MagPathInfo(dataLayer.name, mag.mag, magURI.toString, magURI.toString, hasLocalData = false)
+            } else {
+              val magPath = Paths.get(magURI)
+              val realPath = magPath.toRealPath()
+              // Does this dataset have local data, i.e. the data that is referenced by the mag path is within the dataset directory
+              val isLocal = realPath.startsWith(datasetPath.toAbsolutePath)
+              val unresolvedPath =
+                rawLayerPath.toAbsolutePath.resolve(absoluteLayerPath.relativize(magPath)).normalize()
+              MagPathInfo(dataLayer.name,
+                          mag.mag,
+                          unresolvedPath.toUri.toString,
+                          realPath.toUri.toString,
+                          hasLocalData = isLocal)
+            }
+          }
+        }
+      }
+      case None => List()
+    }
+  }
```

Comment:
This function has grown pretty complex, maybe some of its logic could be extracted to subfunctions? That could make it easier to read. A candidate might be the block handling one layer, and also the block handling one mag.